1-intro/intermediate.ipynb

```diff
@@ -0,0 +1,354 @@
+    "another_set = set([3, 4, 5])\n",
+    "print(my_set)\n",
+    "\n",
+    "my_set.add(4)\n",
+    "print(my_set)\n",
+    "\n",
+    "my_set.remove(2)\n",
+    "print(my_set)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "e32203e8-1b3c-4b5d-a5aa-308a2eb54d2d",
+   "metadata": {},
+   "source": [
+    "Sets support various set operations, such as `union()`, `intersection()`, and `difference()`. These operations allow you to combine and compare sets in useful ways. For example:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "cb11f0af-6deb-4686-ace4-9955192bff28",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "my_set = {1, 2, 3}\n",
+    "\n",
+    "union_set = my_set.union(another_set)\n",
+    "print(union_set)\n",
+    "\n",
+    "intersection_set = my_set.intersection(another_set)\n",
+    "print(intersection_set)\n",
+    "\n",
+    "difference_set = my_set.difference(another_set)\n",
+    "print(difference_set)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "f37af5fd-554e-4df2-9a23-a25ca4832a65",
+   "metadata": {},
+   "source": [
+    "### Lambda functions\n",
+    "Lambda functions are small, anonymous functions that are defined using the `lambda` keyword. They are often used for short, simple operations and can be created on the fly without needing a name. A lambda function can also be assigned to a variable in order to be reusable.\n",
+    "A lambda function has the following basic structure: `lambda arguments: expression`.\n",
+    " - arguments: These are the input parameters that the lambda function takes. You can have zero or more arguments, separated by commas.\n",
+    " - expression: This is the operation or calculation that the lambda function performs using the input arguments.\n",
+    "Lambda functions are typically used for small, single-line operations. For example, you can create a lambda function that adds two numbers:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "e5cca364-0576-4f2c-bb5e-93fb2040a705",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "add = lambda x, y: x + y\n",
+    "print(add(5, 2))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "0c1917c5-bac2-4dd0-94da-37d6935b8413",
+   "metadata": {},
+   "source": [
+    "Here is a slightly more complex example in a snippet of code:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "73b7574e-2bf3-474a-abab-19623c147edc",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Generate some data that you don't know what it will be beforehand\n",
+    "import numpy as np\n",
+    "rng = np.random.default_rng()\n",
+    "multipliers = 10 * rng.random(rng.integers(3, 10, 1))\n",
+    "print(multipliers)\n",
+    "\n",
+    "# Do some operation on the data that is not super simple\n",
+    "complicated_numbers = [(lambda x: (x + 1) / len(multipliers))(num) for num in multipliers]\n",
```

Comment:
The following is exactly equivalent, but much simpler code:
```suggestion
    "complicated_numbers = ((multipliers + 1) / len(multipliers)).tolist()\n",
```
Note that in practice the `.tolist()` is usually not needed, which would make the code even shorted.

Good examples of using lambda functions would be [keys for `sorted()`](https://docs.python.org/3/howto/sorting.html#key-functions). For example, suppose we have some numbers and we want to sort them by their digit sum:
```python
>>> numbers = [19, 100, 123]
>>> sorted_by_digit_sum = sorted(numbers, key=lambda x: sum(int(d) for d in str(x)))
>>> print(sorted_by_digit_sum)
[100, 123, 19]
```

Reply:
I have a hard time coming up with good examples for lambda functions, and I very rarely use them myself. Is it maybe best to drop them completely?

Reply:
Lambda functions are syntactic sugar for passing a single-expression function to some function that accepts other functions as arguments if there is no suitable function defined already. It is always possible to define the function with `def` and then pass it on by name instead of defining it in-place with lambda.

I use lambda functions for [monkeypatching functions in unit tests](https://docs.pytest.org/en/stable/how-to/monkeypatch.html#monkeypatching-functions), but you can see that the `pytest` documentation is using `def` instead. 

In more functionally-oriented languages lambda functions would often be used with `map()` and `filter()` (or with whatever would be equivalent), but in Python using comprehensions is more convenient.

In summary, lambda functions are not useful for the courses, I wouldn't spend time on them.

Reply:
Ok, agreed. I will remove the section on lambda functions.